# EPITOME: Enchanced Phylogenetic Inference Through Optimized Mapping Efficiency

EPITOME attempts to condense a diverse set of DNA sequences (species-level or closer) into discrete, composite sequences that represent the overarching diversity of the dataset. In other words, EPITOME creates sequences which are the _epitome_ of the dataset diversity. This is accomplished by clustering sequencing based on pairwise genetic distances and then summarizing the clusters into a single sequence which contains the most common nucleotide at each position across the aligned cluster sequences (ties selected at random). This can be useful for creating sets of distinct reference genomes for consensus-based viral genome assembly pipelines (like [VAPER](https://github.com/DOH-JDJ0303/VAPER)) which are clustered based on genetic distances that optimize read mapping efficiency.
