# EPITOME: Enchanced Phylogenetic Inference Through Optimized Mapping Efficiency

EPITOME attempts to condense a diverse set of DNA sequences (species-level or closer) into discrete, composite sequences that represent the overarching diversity of the dataset. In other words, EPITOME creates sequences which are the _epitome_ of the dataset diversity. This is accomplished by clustering sequencing into closely related clusters based on their pairwise genetic distances and then summarizing the clusters based on the most common nucleotides observed across the clustered sequences (ties selected at random). This can be useful for creating sets of distinct reference genomes for consensus-based viral genome assembly pipelines (like [VAPER](https://github.com/DOH-JDJ0303/VAPER)) which are clustered based on genetic distances that optimize read mapping efficiency.
